samples/MediatR.Examples.PublishStrategies/Publisher.cs

```diff
@@ -0,0 +1,161 @@
+            PublishStrategies[PublishStrategy.SyncStopOnException] = new CustomMediator(_serviceFactory, SyncStopOnException);
+        }
+
+        public IDictionary<PublishStrategy, IMediator> PublishStrategies = new Dictionary<PublishStrategy, IMediator>();
+        public PublishStrategy DefaultStrategy { get; set; } = PublishStrategy.SyncContinueOnException;
+
+        public Task Publish<TNotification>(TNotification notification)
+        {
+            return Publish(notification, DefaultStrategy, default(CancellationToken));
+        }
+
+        public Task Publish<TNotification>(TNotification notification, PublishStrategy strategy)
+        {
+            return Publish(notification, strategy, default(CancellationToken));
+        }
+
+        public Task Publish<TNotification>(TNotification notification, CancellationToken cancellationToken)
+        {
+            return Publish(notification, DefaultStrategy, cancellationToken);
+        }
+
+        public Task Publish<TNotification>(TNotification notification, PublishStrategy strategy, CancellationToken cancellationToken)
+        {
+            if (!PublishStrategies.TryGetValue(strategy, out var mediator))
+            {
+                throw new ArgumentException($"Unknown strategy: {strategy}");
+            }
+
+            return mediator.Publish(notification, cancellationToken);
+        }
+
+        private Task ParallelWhenAll(IEnumerable<Func<Task>> handlers)
+        {
+            var tasks = new List<Task>();
+
+            foreach (var handler in handlers)
+            {
+                tasks.Add(Task.Run(() => handler()));
+            }
+
+            return Task.WhenAll(tasks);
+        }
+
+        private Task ParallelWhenAny(IEnumerable<Func<Task>> handlers)
+        {
+            var tasks = new List<Task>();
+
+            foreach (var handler in handlers)
+            {
+                tasks.Add(Task.Run(() => handler()));
+            }
+
+            return Task.WhenAny(tasks);
+        }
+
+        private Task ParallelNoWait(IEnumerable<Func<Task>> handlers)
+        {
+            foreach (var handler in handlers)
+            {
+                Task.Run(() => handler());
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private async Task AsyncContinueOnException(IEnumerable<Func<Task>> handlers)
+        {
+            var tasks = new List<Task>();
+            var exceptions = new List<Exception>();
+
+            foreach (var handler in handlers)
+            {
+                try
+                {
+                    tasks.Add(handler());
+                }
+                catch (AggregateException ex)
+                {
+                    exceptions.AddRange(ex.InnerExceptions);
+                }
+                catch (Exception ex)
+                {
+                    exceptions.Add(ex);
+                }
+            }
```

Comment:
Really great examples breakdown - this one particularly seems like a lot of work.  What makes this one AsyncContinueOnException?  Doesn't the ParallelWhenAll behave similarly?

Reply:
AsyncContinueOnException (or just Async, as you can't capture on the 'first' exception) does not run the handler on a new thread. Parallel does, using Task.Run.